app/controllers/locations_controller.rb

```diff
@@ -14,8 +15,144 @@ def index
+    @location.location_code = Location.local_authorities.key(params[:location][:location_admin_district])
+    if @location.save(context: :location_admin_district)
+      if params[:referrer] == "check_answers" || params[:referrer] == "check_local_authority"
+        redirect_to scheme_location_check_answers_path(@scheme, @location, route: params[:route])
+      else
+        redirect_to scheme_location_name_path(@scheme, @location, route: params[:route])
+      end
+    else
+      render :local_authority, status: :unprocessable_entity
+    end
+  end
+
+  def name; end
+
+  def update_name
+    @location.name = params[:location][:name]
+    if @location.save(context: :name)
+      case params[:referrer]
+      when "check_answers"
+        redirect_to scheme_location_check_answers_path(@scheme, @location, route: params[:route])
+      when "details"
+        redirect_to scheme_location_path(@scheme, @location)
+      else
+        redirect_to scheme_location_units_path(@scheme, @location, route: params[:route])
+      end
+    else
+      render :name, status: :unprocessable_entity
+    end
+  end
+
+  def units; end
+
+  def update_units
+    @location.units = params[:location][:units]
+    if @location.save(context: :units)
+      if params[:referrer] == "check_answers"
+        redirect_to scheme_location_check_answers_path(@scheme, @location, route: params[:route])
+      else
+        redirect_to scheme_location_type_of_unit_path(@scheme, @location, route: params[:route])
+      end
+    else
+      render :units, status: :unprocessable_entity
+    end
+  end
+
+  def type_of_unit; end
+
+  def update_type_of_unit
+    @location.type_of_unit = params[:location][:type_of_unit]
+    if @location.save(context: :type_of_unit)
+      if params[:referrer] == "check_answers"
+        redirect_to scheme_location_check_answers_path(@scheme, @location, route: params[:route])
+      else
+        redirect_to scheme_location_mobility_standards_path(@scheme, @location, route: params[:route])
+      end
+    else
+      render :type_of_unit, status: :unprocessable_entity
+    end
+  end
+
+  def mobility_standards; end
+
+  def update_mobility_standards
+    @location.mobility_type = params[:location][:mobility_type]
+    if @location.save(context: :mobility_type)
+      if params[:referrer] == "check_answers"
+        redirect_to scheme_location_check_answers_path(@scheme, @location, route: params[:route])
+      else
+        redirect_to scheme_location_availability_path(@scheme, @location, route: params[:route])
+      end
+    else
+      render :mobility_standards, status: :unprocessable_entity
+    end
+  end
+
+  def availability; end
+
+  def update_availability
+    day = params[:location]["startdate(3i)"]
+    month = params[:location]["startdate(2i)"]
+    year = params[:location]["startdate(1i)"]
+    if [day, month, year].none?(&:blank?)
+      if Date.valid_date?(year.to_i, month.to_i, day.to_i)
```

Comment:
These validations belong on the model. Chat to @asmega

Reply:
Think he's ill today, I'll try to catch him tomorrow